File: 3097. Shortest Subarray With OR at Least K II.java
class Solution {
    public int minimumSubarrayLength(int[] nums, int k) {
        int[] freq = new int[32];
        int answer = Integer.MAX_VALUE, left = 0;
        for (int right = 0; right < nums.length; right++) {
            add(freq, nums[right]);
            while (get(freq) >= k && left <= right) {
                if (answer > right - left + 1) {
                    answer = right - left + 1;
                }
                remove(freq, nums[left]);
                left++;
            }
        }
        return answer == Integer.MAX_VALUE ? -1 : answer;
    }
    private void remove(int[] freq, int num) {
        for (int i = 0; i < 32; i++) {
            if (num % 2 == 1) freq[i] -= 1;
            num /= 2;
        }
    }
    private void add(int[] freq, int num) {
        for (int i = 0; i < 32; i++) {
            if (num % 2 == 1) freq[i] += 1;
            num /= 2;
        }
    }
    private int get(int[] freq) {
        int answer = 0;
        for (int i = freq.length-1; i >= 0; i--) {
            answer *= 2;
            if (freq[i] > 0) {
                answer += 1;
            }
        }
        return answer;
    }
}

File: 3097. Shortest Subarray With OR at Least K II.py
class Solution:
    def minimumSubarrayLength(self, nums: List[int], k: int) -> int:
        def add(num):
            nonlocal freq
            for i in range(32):
                if num % 2 == 1:
                    freq[i] += 1
                num //= 2
        def remove(num):
            nonlocal freq
            for i in range(32):
                if num % 2 == 1:
                    freq[i] -= 1
                num //= 2
        def get():
            nonlocal freq
            ans = 0
            for i in range(31, -1, -1):
                ans *= 2
                if freq[i] > 0:
                    ans += 1
            return ans
        freq = [0] * 32
        answer, left = float('inf'), 0
        for right in range(len(nums)):
            add(nums[right])
            while get() >= k and left <= right:
                answer = min(answer, right - left + 1)
                remove(nums[left])
                left += 1
        return answer if answer != float('inf') else -1
    

